InApp Notifications

swmitra opened the pull request:
This PR adds the required infrastructure to display App notifications created by Brackets publisher with the ability to filter applicability of the notifications based on multiple parameters. A sample notifications file (localized) hosted from S3 would look like - 

```json
{
    "notifications": [
        {
            "sequence": "1555874975721",
            "silent": false,
            "filters": {
                "platforms": ["OSX"],
                "version": "1.14",
                "locales": ["en"],
                "builds": [],
                "extensions": [],
                "lastUsed": ""
            },
            "html": "<div><span>Web development and design trends are changing at a rapid pace. We want to hear how new paradigms and tools are impacting you and the way you work!</span><a id='banner-button' href='https://adobe.ly/bracketsiosurvey'>Take the Brackets Survey</a></div>",
            "actionables": "#banner-button",
            "expiry" : "1556044200000"
        }
    ]
}
```

Commit:
InApp Notification

Commit:
Merge branch 'master' of https://github.com/adobe/brackets into swmitra/Notifications

Commit:
Minor cleanup

Commit:
Remove template parameters

Commit:
MIn height for notification bar

boopeshmahendran commented on src/extensions/default/InAppNotifications/main.js:
```diff
@@ -0,0 +1,372 @@
+            $notificationBarElement = $(NotificationBarHtml);
+
+        // Remove any SCRIPT tag to avoid secuirity issues
+        $notificationBarElement.find('script').remove();
```
Shouldn't we be sanitising `$htmlContent` as this is the network response?

Also, Should we think about more stricter checks for security as there are a lot of other ways to inject scripts in html?  **For Ex:**  Something like `<a onclick="some malicious script">`. 

For reference, https://www.owasp.org/index.php/XSS_Filter_Evasion_Cheat_Sheet

swmitra replied:
As the pre formatted html is provided by us, I don't wanted overwhelming checks and filter here. The checks included here are just to protect against accidental mistake and not from an actual attack.

boopeshmahendran commented on src/extensions/default/InAppNotifications/main.js:
```diff
@@ -0,0 +1,372 @@
+        var result = new $.Deferred();
+        var versionInfoUrl;
+
+        _getNotificationInformation(versionInfoUrl)
```
Why are we calling with `undefined`? Couldn't we just write `_getNotificationInformation()`?

swmitra replied:
Need to add a parameter in the wrapper function which can be provided for testing.

boopeshmahendran commented on src/extensions/default/InAppNotifications/main.js:
```diff
@@ -0,0 +1,372 @@
+                result.reject();
+            });
+
+        return result.promise();
```
nit: I think we can directly return `_getNotificationInformation()` promise instead of creating a newer promise.

swmitra replied:
Need to add a parameter in the wrapper function which can be provided for testing.

boopeshmahendran commented on src/extensions/default/InAppNotifications/main.js:
```diff
@@ -0,0 +1,372 @@
+        $notificationBarElement.find('script').remove();
+
+        // Remove any STYLE tag to avoid styling impact on Brackets DOM
+        $notificationBarElement.find('style').remove();
```
Shouldn't we be sanitising `$htmlContent` as this is the network response?

swmitra replied:
Explanation mentioned before.

boopeshmahendran replied:
Actually, I meant we should be sanitising `$htmlContent` variable instead of the `$notificationBarElement`.

swmitra replied:
Done. Good catch!

boopeshmahendran commented on src/extensions/default/InAppNotifications/main.js:
```diff
@@ -0,0 +1,372 @@
+                            }
+                            // Break, we have acted on one notification already
+                            return false;
+                        }
```
Probably we will need a `return true` here since without it, it will return `undefined` thereby making the `every` to stop iteration. I think `notifications.find` will be more suited to this usecase.

shubhsnov replied:
Nice catch. This is important.

swmitra replied:
Handled.

boopeshmahendran commented on src/extensions/default/InAppNotifications/main.js:
```diff
@@ -0,0 +1,372 @@
+        }
+
+        // PRERELEASE_BEGIN
+        // The following code is needed for supporting notifications in prerelease,
```
Can you please help me understand how does this work only for prerelease?

swmitra replied:
Removed the whole block.

shubhsnov commented on src/extensions/default/InAppNotifications/main.js:
```diff
@@ -0,0 +1,372 @@
+                    locale = locale.substring(0, 2);
+                }
+                switch(locale)  {
+                case "de":
```
This will also work just fine here.
```
locales = ["fr", "de"...];
if (!locales.includes(locale)) {
    locale = "en";
}
```

swmitra replied:
Removed.

boopeshmahendran commented on src/extensions/default/InAppNotifications/main.js:
```diff
@@ -0,0 +1,372 @@
+                    var tmpUrl = _getVersionInfoUrl(brackets.getLocale(), true);
+                    if (tmpUrl !== localVersionInfoUrl) {
+                        $.ajax({
+                            url: tmpUrl,
```
Is there some way to reduce the nesting of promises?

swmitra replied:
Didn't have time to refactor the code, copied from Update notification module.

swmitra replied:
Done. no more "en-US" check first.

swmitra commented on src/extensions/default/InAppNotifications/main.js:
```diff
@@ -0,0 +1,372 @@
+        Strings              = brackets.getModule("strings"),
+        Mustache             = brackets.getModule("thirdparty/mustache/mustache"),
+        HealthLogger         = brackets.getModule("utils/HealthLogger"),
+        NotificationBarHtml  = require("text!htmlContent/notificationContainer.html");
```
I would be using them as I have to add logic to display the notification as a dialog in next commit based on `display` configuration.

swmitra replied:
For now don't have time to bring in Dialogs, removed the imports.

shubhsnov commented on src/extensions/default/InAppNotifications/main.js:
```diff
@@ -0,0 +1,372 @@
+/*
+ * Copyright (c) 2019 - present Adobe Systems Incorporated. All rights reserved.
```
Minor but necessary: You'll need to change this to `Copyright (c) 2019 - present Adobe. All rights reserved.` due to the recent name change for Adobe in all new documents.

swmitra replied:
Done.

shubhsnov commented on src/extensions/default/InAppNotifications/main.js:
```diff
@@ -0,0 +1,372 @@
+        // PRERELEASE_BEGIN
+        // The following code is needed for supporting notifications in prerelease,
+        {
+            if (locale) {
```
This code will run regardless of whether it's a prerelease or normal release. The comment above seems inaccurate.
Also, due to this we will never hit the return statement on line 96.

swmitra replied:
Removed.

shubhsnov reviewed (commented):
Apart from a few issues that I and Boopesh have mentioned this looks functionally solid to me.

swmitra commented on src/extensions/default/InAppNotifications/main.js:
```diff
@@ -0,0 +1,372 @@
+        }
+
+        // If more than 24 hours have passed since our last fetch, fetch again
+        if ((new Date()).getTime() > lastInfoURLFetchTime + ONE_DAY) {
```
Done.

swmitra commented on src/extensions/default/InAppNotifications/htmlContent/notificationContainer.html:
```diff
@@ -0,0 +1,7 @@
+    <div id="content-container">
+    </div>
+    <div id="close-icon-container">
+        <button type="button" id="close-icon" tabIndex="0">&times;</button>
```
Changed to `.` selector.

swmitra commented on src/brackets.config.prerelease.json:
```diff
@@ -4,5 +4,6 @@
     "serviceKey"              : "brackets-service",
     "environment"             : "production",
     "update_info_url"         : "https://s3.amazonaws.com/files.brackets.io/updates/prerelease/<locale>.json",
+    "notification_info_url"         : "https://s3.amazonaws.com/files.brackets.io/notifications/prerelease/<locale>.json",
```
Fixed.

swmitra commented on src/extensions/default/InAppNotifications/styles/styles.css:
```diff
@@ -0,0 +1,56 @@
+    left: 0px;
+    bottom: 25px;
+    outline: none;
+	overflow: hidden;
```
Done.

Commit:
Address code review comments

swmitra commented:
@boopeshmahendran and @shubhsnov Thanks a ton for reviewing this PR minutely under short notice 💯 
Addressed all comments now. 
@shubhsnov Please test it once more whenever possible.

boopeshmahendran commented on src/extensions/default/InAppNotifications/styles/styles.css:
```diff
@@ -0,0 +1,56 @@
+    float: right;
+    text-align: center;
+    width: auto;
+	min-width: 66px;
```
nit: some fomatting issues in this file.

shubhsnov commented:
LGTM 👍 Tested with multiple platforms, versions etc. Everything is working fine. With this, we can target notifications in the future. Thanks, @swmitra for this brilliant contribution.
